barretenberg/cpp/src/barretenberg/stdlib/primitives/databus/databus.cpp

```diff
@@ -0,0 +1,62 @@
+#include "databus.hpp"
+#include "../circuit_builders/circuit_builders.hpp"
+
+namespace bb::stdlib {
+
+template <typename Builder> void databus<Builder>::bus_vector::set_values(const std::vector<field_pt>& entries_in)
+{
+    raw_entries = entries_in;
+    length = raw_entries.size();
+}
+
+template <typename Builder> void databus<Builder>::bus_vector::initialize() const
+{
+    ASSERT(!initialized);
+    ASSERT(context != nullptr);
+    // Populate the bus vector entries from raw entries
+    for (const auto& entry : raw_entries) {
+        if (entry.is_constant()) { // create a non-constant witness from the constant witness
+            auto const_var_idx = context->put_constant_variable(entry.get_value());
+            entries.emplace_back(field_pt::from_witness_index(context, const_var_idx));
+        } else { // normalize the raw entry
+            entries.emplace_back(entry.normalize());
+        }
+        // Add the entry to the bus vector data
+        context->append_to_bus_vector(bus_idx, entries.back().get_witness_index());
+    }
+
+    initialized = true;
+}
+
+template <typename Builder> field_t<Builder> databus<Builder>::bus_vector::operator[](const size_t index) const
+{
+    if (index >= length) {
+        ASSERT(context != nullptr);
+        context->failure("bus_vector: access out of bounds");
+    }
+    return entries[index];
+}
+
+template <typename Builder> field_t<Builder> databus<Builder>::bus_vector::operator[](const field_pt& index) const
+{
+    auto raw_index = static_cast<size_t>(uint256_t(index.get_value()).data[0]);
```

Comment:
Let's discuss this constant code path. I based the design of this class after the ROM table class which is why it's here. Originally thought it made sense but I'm not sure there's a valid reason to be able to extract databus entries without creating a gate.

Reply:
Wouldn't the prover be able to replace the value in the witness, since there is nothing connecting the variable to databus column?